### ISSUE TITLE
Persistent subscription with StartFromBeginning option resends messages after EventStore restart

### ISSUE THREAD

**Issue by gaevoy** [EventStore/EventStore#1081]
After couple of server restarts due to Windows Update I noticed that old messages (saved and processed already 1 month ago) are resent again via PersistentSubscription and _StartFromBeginning_ option. I aware that you use at-least-one semantic for PersistentSubscription, but it is not reproduced if your subscription _StartFromCurrent_. See test below to reproduce the bug.
We run EventStore using following command: `EventStore.ClusterNode.exe --db ./db --log ./logs`.
Environment info:

```
ES VERSION:               3.9.0.0 (oss-v3.9.0/ec9208bd5751646cf7d12dcc9d98b30045714175, Sat, 27 Aug 2016 11:30:03 +0100)
OS:                       Windows (Microsoft Windows NT 6.2.9200.0)
RUNTIME:                  .NET 4.0.30319.42000 (64-bit)
EventStore.Client         3.8.0.
```

Run test with debugger attached, once you paused on breakpoint restart EventStore and then resume test execution.

```
using System;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using EventStore.ClientAPI;
using NUnit.Framework;

[TestFixture]
public class PersistentSubscriptionWithStartFromBeginningBug
{
    [Test]
    public async Task ShouldNotResendMessagesAfterRestart()
    {
        // Given
        var stream = "PersistentSubscriptionTest" + Guid.NewGuid();
        var group = "test";
        var processedMessagesBefore = 0;
        using (var con = await OpenConnection())
        {
            var settings = PersistentSubscriptionSettings.Create()
                .StartFromBeginning(); // <--- If you remove StartFromBeginning bug disappears
            await con.CreatePersistentSubscriptionAsync(stream, group, settings, null);
            var subscription = await con.ConnectToPersistentSubscriptionAsync(stream, group, (s, e) => { processedMessagesBefore++; }, (s, r, e) => { }, bufferSize: 1);
            await con.AppendToStreamAsync(stream, ExpectedVersion.Any,
                new EventData(Guid.NewGuid(), "TestEvent1", false, null, null),
                new EventData(Guid.NewGuid(), "TestEvent2", false, null, null),
                new EventData(Guid.NewGuid(), "TestEvent3", false, null, null));
            await Task.Delay(5000); // Wait while all messages acked
            subscription.Stop(Timeout.InfiniteTimeSpan);
        }

        // When
        Debugger.Break(); // <--- Restart EventStore here and wait when it started!!!
        var processedMessagesAfter = 0;
        using (var con = await OpenConnection())
        {
            await con.ConnectToPersistentSubscriptionAsync(stream, group, (s, e) => { processedMessagesAfter++; }, (s, r, e) => { }, bufferSize: 1);
            await Task.Delay(5000); // Wait for some messages
        }

        // Then
        Assert.AreEqual(3, processedMessagesBefore);
        Assert.AreEqual(0, processedMessagesAfter);
    }

    private static async Task<IEventStoreConnection> OpenConnection()
    {
        var connection = EventStoreConnection.Create("ConnectTo=tcp://admin:changeit@localhost:1113");
        await connection.ConnectAsync();
        return connection;
    }
}
```

**Comment by hayley-jean** [comment]
Hi @gaevoy 

You are most likely seeing this because your persistent subscription is not getting enough events to create a checkpoint.

When you restart Event Store, your persistent subscription will continue from its last checkpoint.
If there is none, it will start from 0 again.

A checkpoint will be written when either:
 - The maximum number of events are received (specified by `MaximumCheckPointCountOf`)
 - The minimum number of events are received and the minimum amount of time has elapsed (specified by `MinimumCheckPointCountOf` and `CheckpointAfter`)

In your case, you are using the default settings, so at the very least your subscription will only checkpoint after receiving 10 events and waiting 2 seconds. This means that 3 events will never cause a checkpoint to be created.

For this test, setting `MinimumCheckPointCountOf` to 1 would allow your subscription to create a checkpoint :
```
PersistentSubscriptionSettings.Create()
                 .MinimumCheckPointCountOf(1)
                 .StartFromBeginning();
```
 
_However_, this test will still fail on the current version of Event Store.

There is an existing issue with persistent subscriptions where the checkpointed event will be re-sent to your subscribers after Event Store is restarted. So instead of seeing all of the events being re-sent, you will only see 1.

This has been fixed by https://github.com/EventStore/EventStore/pull/1075, but it is not yet released.

**Comment by gaevoy** [comment]
@hayley-jean thank you for so detailed explanation. 
I'm still don't understand why it is not reproduced if I use *StartFromCurrent* option (simply replace *.StartFromBeginning()* with *.StartFromCurrent()* in test above). 
Can you please clarify?

**Comment by gregoryyoung** [comment]
@gaevoy Because StartFromCurrent says to start from where ever the stream is in absence of a checkpoint.

What is happening here is you write and process 3 events. Your checkpoint is set to checkpoint after 10 so no checkpoint is written.

When you restart the node and recreate the subscription there is no checkpoint present. As such it falls back on what you tell it to do in the subscription arguments. StartFromBeginning will start at event 0. StartFromCurrent will start from event 3 (current).

If you set a lower value for checkpointing in your config so it writes a checkpoint it will then start at the checkpointed position not at the default position on the subscription.

**Comment by pgermishuys** [comment]
Fixed by #1075
